File: Python/examples/varlen_channel_inprocess.py
from __future__ import annotations

import argparse
import os
import time

from _common import cleanup_shm, load_xproc, wait_for_message


MESSAGES = ["hello", "xproc", "variable-length", "messages"]


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--module-dir")
    args = parser.parse_args()

    xproc = load_xproc(args.module_dir)

    shm_path = f"/xproc_python_varlen_inprocess_{os.getpid()}"
    cleanup_shm(xproc, shm_path)

    create_options = xproc.TransportOptions()
    create_options.path = shm_path
    create_options.shm_size = xproc.shm_size_for_data_capacity(32_768)
    create_options.channel_type = xproc.ChannelType.VARLEN
    create_options.create_if_missing = True

    attach_options = xproc.TransportOptions()
    attach_options.path = shm_path
    attach_options.shm_size = xproc.INFER_EXISTING_SHM_SIZE
    attach_options.channel_type = xproc.ChannelType.VARLEN
    attach_options.create_if_missing = False

    producer = xproc.Producer(create_options)
    consumer = xproc.Consumer(attach_options)

    received = 0

    try:
        for message in MESSAGES:
            producer.send_varlen(message)
            time.sleep(0.008)

            payload = wait_for_message(consumer.poll_copy, label="varlen payload")
            text = payload.decode("utf-8")
            print(f"recv: {text}")
            if text != MESSAGES[received]:
                raise RuntimeError(f"sequence mismatch, expected {MESSAGES[received]} got {text}")
            received += 1
    finally:
        consumer.close()
        producer.close()
        cleanup_shm(xproc, shm_path)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: node/examples/varlen_channel_inprocess.ts
const timers = require("node:timers/promises");

const xproc = require("../index.js") as XprocModule;

const delay = timers.setTimeout as (ms: number) => Promise<void>;

function cleanupShm(path: string): void {
  try {
    xproc.shmUnlink(path);
  } catch {
    // Best-effort cleanup only.
  }
}

async function main(): Promise<void> {
  const shmPath = `/xproc_node_varlen_inprocess_${process.pid}`;
  cleanupShm(shmPath);

  const messages = ["hello", "xproc", "variable-length", "messages"];
  const created = xproc.shm.createVarlenChannel({
    path: shmPath,
    dataCapacity: 32768n,
  });
  const producer = created.openProducer();
  const consumer = xproc.shm.attachVarlenChannel({ path: shmPath }).openConsumer();

  let received = 0;

  try {
    for (const message of messages) {
      producer.sendVarlen(message);
      await delay(8);

      while (true) {
        const payload = consumer.pollCopy();
        if (payload === null) {
          await delay(1);
          continue;
        }

        const text = Buffer.from(payload).toString("utf8");
        console.log(`recv: ${text}`);
        if (text !== messages[received]) {
          throw new Error(`sequence mismatch, expected ${messages[received]} got ${text}`);
        }
        received += 1;
        break;
      }
    }
  } finally {
    consumer.close();
    producer.close();
    cleanupShm(shmPath);
  }
}

void main().catch((error: unknown) => {
  const message = error instanceof Error ? error.stack ?? error.message : String(error);
  console.error(message);
  process.exit(1);
});
